db/api_history.c

```diff
@@ -0,0 +1,173 @@
+/*
+   Copyright 2024 Bloomberg Finance L.P.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+ */
+
+#include <assert.h>
+#include <ctype.h>
+#include <locks_wrap.h>
+#include <plhash.h>
+#include <pthread.h>
+#include <string.h>
+
+#include "logmsg.h"
+
+#include "api_history.h"
+
+struct api_history {
+    hash_t *entries;
+    pthread_rwlock_t lock;
+};
+
+static int hash_entry(api_driver_t *entry, int len)
+{
+    return hash_default_fixedwidth((unsigned char*)entry->name, strlen(entry->name));
+}
+
+static int compare_entries(api_driver_t *entry1, api_driver_t *entry2, int len)
+{
+    int rc = strcmp(entry1->name, entry2->name);
+    if (!rc) rc = strcmp(entry1->version, entry2->version);
+    return rc;
+}
+
+static char *toLower(char *str)
+{
+    char *curr = str;
+    
+    while (*curr) {
+        *curr = tolower(*curr);
+        ++curr;
+    }
+    
+    return str;
+}
+
+static api_driver_t *create_entry(char *api_driver_name, char *api_driver_version)
+{
+    const char default_value[] = "unknown";
+    api_driver_t *entry = malloc(sizeof(api_driver_t));
+
+    entry->name = api_driver_name ? toLower(strdup(api_driver_name)) : strdup(default_value);
+    entry->version = api_driver_version ? toLower(strdup(api_driver_version)) : strdup(default_value);
+
+    return entry;
+}
+
+static int free_entry(api_driver_t *entry, void *arg)
+{
+    free(entry->name);
+    free(entry->version);
```

Comment:
does `entry` itself needs freeing?

Reply:
`entry` is freed in `update_api_history`. I updated `free_entry` that will now accept a flag for setting whether the pointer should also be freed.